# Helper workflow to run LDSC (linkage score disequilibrium on all aligned gwas summary stats)

## Required inputs
This is intended to be run on all aligned gwas summary stats produced by gwasqc. The required inputs are as follows:
1) Map file (example provided in example_input/map_file.tsv)
2) RSID Map file. This can be generated by download the the VCF here, and then creating a file with two columns, the first being the snp ID (chr:pos:ref:alt) and the second being the rsid, no header needed. 
3) HapMap snplist, and ldsc reference file (these can be downloaded here: https://ibg.colorado.edu/cdrom2021/Day06-nivard/GenomicSEM_practical/eur_w_ld_chr/, and https://github.com/perslab/CELLECT/blob/master/data/ldsc/w_hm3.snplist)
4) In the config file make sure that there is a population prevalence rate that aligns with the diseases listed in the phenotype colum nof the map file.

## Outputs
Each input will generate an output log file which will contain the LDSC results.

## Environment Build & Running the Workflow
Please note there is no example data. Please see [LDSC](https://github.com/bulik/ldsc.git) for details. This workflow simply runs LDSC on all aligned summary stats. 

** You must add the LDSC output to meta data to successfully run [run-metal](https://github.com/pozdeyevlab/gwas-analysis/tree/main/run-metal)**
```bash
# Clone LDSC
git clone https://github.com/bulik/ldsc.git

conda env create -f environment.yml
conda activate run_ldsc
poetry install
snakemake --cores 1 --configfile config.yaml --use-conda --conda-frontend conda
```


